app/Services/ProvisioningService.php

```diff
@@ -0,0 +1,366 @@
+        'permissions.roomTypes.*' => 'string',
+        'permissions.servers' => 'list',
+        'permissions.servers.*' => 'string',
+        'permissions.serverPools' => 'list',
+        'permissions.serverPools.*' => 'string',
+    ];
+
+    public function create(object $properties)
+    {
+        $this->createWrapper($properties, function ($role) use ($properties) {
+            foreach ($properties->permissions as $group => $perms) {
+                foreach ($perms as $item) {
+                    $permName = "$group.$item";
+                    $perm = Permission::firstWhere('name', $permName);
+                    if (is_null($perm)) {
+                        throw new RecordsNotFoundException("Could not find permission with name '$permName'");
+                    }
+                    $permissions[] = $perm->id;
+                }
+            }
+            $role->name = $properties->name;
+            $role->save();
+            $role->permissions()->sync($permissions);
+        });
+    }
+
+    public function destroy(array $match = [])
+    {
+        $this->destroyWrapper($match);
+    }
+}
+
+class UserProvisioner extends AbstractProvisioner
+{
+    protected string $model = User::class;
+
+    protected array $expectedProperties = [
+        'firstname' => 'required|string',
+        'lastname' => 'required|string',
+        'email' => 'required|string',
+        'password' => 'required|string',
+        'authenticator' => 'required|string',
+        'roles' => 'required|list|exists:roles,name',
+        'roles.*' => 'string',
+        'locale' => 'required|string',
+        'timezone' => 'required|string',
+    ];
+
+    protected function instanceName(object $properties)
+    {
+        return "$properties->firstname $properties->lastname";
+    }
+
+    public function create(object $properties)
+    {
+        $this->createWrapper($properties, function ($user) use ($properties) {
+            $roles = Role::whereIn('name', $properties->roles)->get();
+            if (count($properties->roles) != count($roles)) {
+                $message = "Could not find specified role(s) for user '$properties->firstname $properties->lastname'";
+                Log::error($message);
+                throw new RecordsNotFoundException($message);
+            }
+            $user->firstname = $properties->firstname;
+            $user->lastname = $properties->lastname;
+            $user->email = $properties->email;
+            $user->password = \Hash::make($properties->password);
+            $user->authenticator = $properties->authenticator;
+            $user->locale = $properties->locale;
+            $user->timezone = $properties->timezone;
+            $user->save();
+            $user->roles()->sync($roles);
+        });
+    }
+
+    public function destroy(array $match = [])
+    {
+        $this->destroyWrapper($match);
+    }
+}
+
+class SettingsProvisioner
+{
+    public static array $expectedProperties = [
```

Comment:
This is far less strict than app/Http/Requests/UpdateSettings.php, unsure if this can cause trouble if values are set out of the desired range

Reply:
Done. Also found bugs in test data :face_with_peeking_eye:.